### ISSUE TITLE
Deletes are not propagated to BatchHandler.m_rowKeyCache

### ISSUE THREAD

**Issue by gabrielreid** [kairosdb/kairosdb#575]
When performing a delete of a range of data points, the delete is not propagated to the row key cache in BatchHandler. This has as effect that writing new data for the same metric with the same tags after the delete will not write a new entry in the row key table.

This situation can occur if historical data is ingested into KairosDB, then a data issue is discovered (e.g. numerical data was imported as strings), so the data is deleted and then reimported. In this case, all data points will be imported but the newly-imported data won't be represented in the row key table, meaning that the re-imported data will not show up in queries.

One simple (but not fully sufficient) way of resolving this is to clear the corresponding entries in `BatchHandler.m_rowKeyCache` when performing a delete, although this won't be sufficient if there are multiple instances of KairosDB running.

Another option (although also sub-optimal) would be to not delete the row key entry at all for deletes -- however, this feels like optimizing for delete correctness at the cost of query performance.

**Comment by brianhks** [comment]
What version of Kairos are you using?  I've had this problem in the past and the code currently will purge the cache when a delete operation is sent.  But you bring up an interesting situation of multiple Kairos nodes.  The real problem is if you have separated ingest from query nodes, hmm not sure how to solve this.

**Comment by gabrielreid** [comment]
We're encountering this on an internal fork of 1.2.1. However, looking at the `develop` branch, I don't see the cache being purged with a delete operation. Adding the following test to `CassandraDatastoreTest` seems to confirm that:

```
	@Test
	public void test_deleteDataPoints_DeleteThenWriteThenRead() throws IOException, DatastoreException, InterruptedException
	{
		String metricName = "DeleteThenWritex";
		long rowKeyTime = CassandraDatastore.calculateRowTime(s_dataPointTime);

		DataPointSet dpSet = new DataPointSet(metricName);
		dpSet.addDataPoint(new LongDataPoint(rowKeyTime, 1));
		putDataPoints(dpSet);
		Thread.sleep(1000);

		DatastoreMetricQuery query = new DatastoreMetricQueryImpl(metricName, EMPTY_MAP, Long.MIN_VALUE, Long.MAX_VALUE);

		CachedSearchResult res = createCache(metricName);
		s_datastore.queryDatabase(query, res);
		List<DataPointRow> rows = res.getRows();
		assertThat(rows.size(), equalTo(1));

		s_datastore.deleteDataPoints(query);
		Thread.sleep(2000);

		dpSet = new DataPointSet(metricName);
		dpSet.addDataPoint(new LongDataPoint(rowKeyTime + 1000, 2));
		putDataPoints(dpSet);
		Thread.sleep(1000);

		// This last check currently fails because the row key cache still contains the deleted row key
		res = createCache(metricName);
		s_datastore.queryDatabase(query, res);
		rows = res.getRows();
		assertThat(rows.size(), equalTo(1));

	}
```

It's definitely not a simple issue to resolve if you take multiple KairosDB nodes into account -- our current internal approach is mostly just to document the delete behaviour to help manage expectations. Making the scheduling of the `CleanRowKeyCache` class configurable (it's currently hard-coded to 1 hour) would also probably help, but it definitely still doesn't get things to the point where you can just delete some incorrectly-ingested data and then re-ingest it.

**Comment by brianhks** [comment]
Right here: https://github.com/kairosdb/kairosdb/blob/develop/src/main/java/org/kairosdb/datastore/cassandra/CassandraDatastore.java#L844
there is a flag that if get set (when an entire 3 week row is removed) it clears the cache at the end of the operation.

**Comment by gabrielreid** [comment]
Aha, thanks for pointing that out @brianhks, my mistake -- I didn't realize that the row key cache was being injected into `CassandraDataStore` and was therefore a shared instance with `BatchHandler`, so I missed this. The construction of the CassandraDataStore is done manually in `CassandraDatastoreTest` and doesn't inject the same row key cache in the same way, which is why my test case above fails.

It turns out that the reason that we're running into this is indeed due to having multiple Kairos nodes that handle updates -- when one node handles a delete and then another node handles a following write, the row key cache is out of sync and results in the problem logged here.

I'm thinking that resolving this is probably outside the scope of Kairosdb, what do you think?

**Comment by brianhks** [comment]
Yes and no.  Eventually Kairos will need to work together as a cluster for various reasons.  And when they do they should notify each other of when the cache needs to be rebuilt.  There may be a way to do it today by using Cassandra, I'll have to think about it.

**Comment by gabrielreid** [comment]
If the intention is to have support writes distributed over multiple KairosDB nodes, while also _reliably_ supporting the pattern of `write -> delete -> write` without any data loss, then I think that the approach with the most chances of success are to drop the row key cache entirely (or at least limit its scope to a single batch write).

Doing this would obviously result in a lot more writes going to Cassandra, but I think that this would also allow the memstore in Cassandra to play basically the same role that the row key cache currently plays, in that the memstore would deduplicate the writes to the row key table before they are flushed to SSTables. The write-ahead log in Cassandra would still need to take the overhead of the duplicate writes, which would be an additional load while writing as well as during recovery, but I would expect that this overhead would not be overly noticeable in write performance of KairosDB.

The alternate options that I can imagine (maintain the row key cache and coordinate it over multiple nodes, or perform the the equivalent of checkAndPut on writes to the row key table) would likely either result in much higher latency on writes, or not reliably resolve the issue (i.e. would mean that it's not totally certain that there will be no data loss if `write -> delete -> write` is performed without any extra waiting between steps).

Any ideas how much of a write performance boost the use of the row key cache offers currently?

**Comment by brianhks** [comment]
The write cache has a big benefit.  You can effectively try turning it off by setting it to a very low value.  It will double the traffic to C*
